mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+        match message {
+            (socket, Ok(stream)) => {
+                let task_internal_tx = self.internal_msg_tx.clone();
+                let Some(remote_socket) = self.mappings.get(&socket).cloned() else {
+                    return Err(PortForwardError::SocketMappingNotFound(socket));
+                };
+                let (response_tx, mut response_rx) = mpsc::channel(16);
+                self.tasks.insert(remote_socket, response_tx);
+                tokio::spawn(async move {
+                    let (read, mut write) = stream.into_split();
+                    let mut read_stream = ReaderStream::new(read);
+                    let (oneshot_tx, oneshot_rx) = oneshot::channel::<ConnectionId>();
+                    let _ = task_internal_tx
+                        .send(PortForwardMessage::Connect(remote_socket, oneshot_tx))
+                        .await;
+                    let Ok(connection_id) = oneshot_rx.await else {
+                        return Err(PortForwardError::InternalMessageError(format!("Error while trying to receive connection ID for socket address {socket}")));
+                    };
+
+                    loop {
+                        select! {
+                            message = read_stream.next() => match message {
+                                Some(Ok(message)) => {
+                                    let _ = task_internal_tx.send(PortForwardMessage::Send(connection_id, message.into())).await;
+                                },
+                                Some(Err(error)) => {
+                                    break Err(PortForwardError::TcpListenerError(error));
+                                },
+                                None => {
+                                    let _ = task_internal_tx.send(PortForwardMessage::Close(remote_socket, connection_id)).await;
+                                    break Ok(());
+                                },
+                            },
+
+                            message = response_rx.recv() => match message {
+                                Some(message) => {
+                                    match write.write(message.as_ref()).await {
+                                        Ok(_) => continue,
+                                        Err(error) => break Err(PortForwardError::InternalMessageError(format!("Failed to write data to local port in task: {error}"))),
+                                    }
+                                },
+                                None => break Ok(()), // PortForwarder received DaemonTcpOutgoing::Close
+                            }
+                        }
+                    }
+                });
+            }
+            (socket, Err(error)) => {
+                // error from TcpStream
+                tracing::error!("Error occured while listening to local socket {socket}: {error}");
+                self.listeners.remove(&socket);
+            }
```

Comment:
If you started this whole function with
```rust
let stream = match message.1 {
    Ok(stream) => stream,
    Err(error) => {
        // log and remove listener
        return Ok(());
    }
};
```
it would flatten the code a bit :V